File: internal/gateway/join_test.go
package gateway_test

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"path/filepath"
	"strings"
	"testing"

	"github.com/PatrikLager/vtt-platform/internal/campaign"
	"github.com/PatrikLager/vtt-platform/internal/gateway"
	"github.com/PatrikLager/vtt-platform/internal/identity"
)

// The shared join link (joining-a-table spec §2, plan J2).
//
// This endpoint is UNAUTHENTICATED BY CONSTRUCTION — that is its whole point —
// so it is the one surface in the product where getting a refusal wrong hands
// something to a stranger. It mints a spectator and returns a token, and it
// does nothing else: no campaign state, no event, no role from the caller.

type joinFixture struct {
	t   *testing.T
	srv *httptest.Server
	ids *identity.DB

	// path is the campaign file, so a test can count ROWS. The endpoint's
	// refusal properties are about what it did NOT create, and a decoded
	// reply cannot witness that — see post's comment for how that went wrong.
	path string
}

func newJoinFixture(t *testing.T) *joinFixture {
	t.Helper()
	path := filepath.Join(t.TempDir(), "campaign.db")
	c, err := campaign.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { c.Close() })
	ids, err := identity.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { ids.Close() })

	srv := httptest.NewServer(gateway.New(c, ids).Handler())
	t.Cleanup(srv.Close)
	return &joinFixture{t: t, srv: srv, ids: ids, path: path}
}

// count returns the number of rows in table, straight from the file.
func (f *joinFixture) count(table string) int {
	f.t.Helper()
	raw, err := sql.Open("sqlite", f.path)
	if err != nil {
		f.t.Fatal(err)
	}
	defer raw.Close()
	var n int
	if err := raw.QueryRow(`SELECT count(*) FROM ` + table).Scan(&n); err != nil {
		f.t.Fatal(err)
	}
	return n
}

type joinReply struct {
	Token         string `json:"token"`
	ParticipantID string `json:"participantId"`
	Name          string `json:"name"`
	Role          string `json:"role"`
}

// post returns the response, the decoded reply, AND THE RAW BODY.
//
// The raw body is not a convenience. The first version of this fixture
// compared decoded joinReply structs to check that two refusals were
// identical — but http.Error writes PLAIN TEXT, so decoding failed and both
// sides were zero values, identical whatever the bodies actually said. The
// central security property of this endpoint was pinned by nothing, and
// injection proved it: making the closed-door case say so distinctly failed
// no test.
func (f *joinFixture) post(secret, name string) (*http.Response, joinReply, string) {
	f.t.Helper()
	body, err := json.Marshal(map[string]string{"secret": secret, "displayName": name})
	if err != nil {
		f.t.Fatal(err)
	}
	resp, err := http.Post(f.srv.URL+"/join", "application/json", strings.NewReader(string(body)))
	if err != nil {
		f.t.Fatal(err)
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		f.t.Fatal(err)
	}
	var out joinReply
	_ = json.Unmarshal(raw, &out)
	return resp, out, string(raw)
}

func TestJoiningThroughAnOpenDoorMintsASpectator(t *testing.T) {
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}

	resp, got, _ := f.post(secret, "Kim")
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("status = %d, want 200", resp.StatusCode)
	}
	if got.Token == "" || got.ParticipantID == "" {
		t.Fatalf("a join must return a usable credential, got %+v", got)
	}
	// SPECTATOR, always. The caller does not choose (spec §2, §5).
	if got.Role != string(identity.RoleSpectator) {
		t.Fatalf("role = %q, want spectator — anyone through the door must be unable to act",
			got.Role)
	}
	if got.Name != "Kim" {
		t.Fatalf("name = %q, want the name they gave", got.Name)
	}

	// And the credential is REAL: it verifies to the participant just minted.
	p, err := f.ids.Verify(got.Token)
	if err != nil {
		t.Fatalf("the returned token must verify: %v", err)
	}
	if p.ID != got.ParticipantID || p.Role != identity.RoleSpectator {
		t.Fatalf("verified %+v, want the spectator that was returned", p)
	}
}

func TestAClosedDoorAndAWrongSecretAreRefusedIDENTICALLY(t *testing.T) {
	// The security property of this endpoint. A distinguishable refusal tells
	// a prober which half they got right — whether the campaign exists and is
	// merely shut, or whether their guess at the secret was wrong. Verify()
	// already sets this precedent deliberately for tokens.
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}

	// Door CLOSED, secret RIGHT.
	closedRight, _, bodyA := f.post(secret, "Kim")
	// Door OPEN, secret WRONG.
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}
	openWrong, _, bodyB := f.post("not-the-secret", "Kim")

	if closedRight.StatusCode != openWrong.StatusCode {
		t.Fatalf("status differs: closed-door=%d wrong-secret=%d — the difference tells a "+
			"prober which half they got right", closedRight.StatusCode, openWrong.StatusCode)
	}
	// RAW bodies, not decoded structs: these responses are plain text, so a
	// decoded comparison sees two zero values and can never fail.
	if bodyA != bodyB {
		t.Fatalf("body differs:\n  closed door: %q\n  wrong secret: %q\nthat difference "+
			"tells a prober which half they got right", bodyA, bodyB)
	}
	if closedRight.StatusCode == http.StatusOK {
		t.Fatal("neither of those may succeed")
	}
}

func TestAClosedDoorMintsNobody(t *testing.T) {
	// Not just refused — NO ROW. A refusal that still created a participant
	// would make this unauthenticated endpoint a way for any stranger to fill
	// the table's database.
	//
	// THE FIRST VERSION OF THIS TEST COULD NOT FAIL, and it is the same defect
	// post's comment describes one screen above. It asserted on the DECODED
	// reply, but http.Error writes plain text: got.Token was always "", so the
	// first assertion was a tautology and the second was Verify(""), which
	// errors unconditionally. Injection proved it — minting a participant on
	// every refused join failed nothing in this package OR in identity. The
	// rows are the only honest witness to a claim about what was not created.
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	before := f.count("participants")
	resp, _, _ := f.post(secret, "Kim")
	if resp.StatusCode == http.StatusOK {
		t.Fatal("a closed door must refuse even the correct secret")
	}
	if after := f.count("participants"); after != before {
		t.Fatalf("a refused join minted %d participant(s) anyway", after-before)
	}
}

func TestARefusedJoinWritesNothingAtAll(t *testing.T) {
	// Spec §2 rests its case against rate limiting entirely on the closed door
	// leaving "no standing endpoint to hammer" — the link is INERT. That is a
	// claim about writes, not just about credentials, and it has to be checked
	// HERE rather than only in identity: identity.JoinAllows can be perfectly
	// read-only while this handler still reaches the minting path. That seam
	// is precisely the shape this plan was built around.
	//
	// Nothing below asks for the secret first: JoinSecret() MINTS the row, so
	// a fixture that warms it up cannot see this.
	f := newJoinFixture(t)
	if n := f.count("join_access"); n != 0 {
		t.Fatalf("fixture is not fresh: %d join_access row(s) before any request", n)
	}
	resp, _, _ := f.post("a stranger's guess", "Kim")
	if resp.StatusCode == http.StatusOK {
		t.Fatal("a campaign whose door was never opened must refuse")
	}
	if n := f.count("join_access"); n != 0 {
		t.Fatalf("an anonymous, REFUSED request wrote %d row(s) — it takes SQLite's write "+
			"lock on the file internal/store appends every event to, on the one path a "+
			"stranger controls", n)
	}
}

func TestAnOversizedBodyIsRefusedBeforeItIsRead(t *testing.T) {
	// The cap on the ONE surface an anonymous stranger controls, and it was
	// exercised by nothing: raise maxJoinBody to a gigabyte and everything
	// still passed.
	//
	// The oversize goes in the SECRET, not the display name. An 8KB name is
	// refused 400 by the rune bound either way, so that shape would pass
	// whether the cap existed or not — the wrong-reason trap, in the test
	// written to close a gap.
	f := newJoinFixture(t)
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}
	before := f.count("participants")

	resp, _, body := f.post(strings.Repeat("x", 8<<10), "Kim")

	if resp.StatusCode != http.StatusBadRequest {
		t.Fatalf("status = %d, want %d for a body past the cap", resp.StatusCode, http.StatusBadRequest)
	}
	// Refused as MALFORMED, not as a door problem: the request never got far
	// enough to be a guess at the secret, and reporting it as one would tell a
	// prober their oversized body was merely the wrong password.
	if !strings.Contains(body, "malformed") {
		t.Fatalf("an oversized body must be refused as malformed, got %q", strings.TrimSpace(body))
	}
	if after := f.count("participants"); after != before {
		t.Fatal("an oversized request minted a participant")
	}
}

func TestAnEmptyDisplayNameIsRefused(t *testing.T) {
	// It is what the whole table sees. Blank, or whitespace pretending to be
	// blank, is not a name.
	//
	// Refused DISTINCTLY, and asserted as such: this is the joiner's own
	// mistake and saying so leaks nothing about the door or the secret. The
	// earlier version asserted only "not 200", so collapsing this into the
	// shared refusal failed nothing — and that would tell somebody who simply
	// forgot to type their name that the link is closed.
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}
	for _, name := range []string{"", "   ", "\t\n"} {
		resp, _, body := f.post(secret, name)
		if resp.StatusCode != http.StatusBadRequest {
			t.Fatalf("display name %q: status %d, want %d", name, resp.StatusCode, http.StatusBadRequest)
		}
		if strings.Contains(body, "not accepting anyone") {
			t.Fatalf("display name %q was refused with the DOOR's message (%q) — a joiner who "+
				"forgot their name would be told the link is closed", name, strings.TrimSpace(body))
		}
	}
}

func TestADisplayNameIsBoundedAndPrintable(t *testing.T) {
	// An UNAUTHENTICATED caller chooses this string, and every client at the
	// table then renders it in every presence frame. Length and control
	// characters are two different ways for a stranger to decide what everyone
	// else's screen does — the client escapes with textContent, so this is not
	// XSS, but "not XSS" is not the same as "bounded".
	//
	// Refused distinctly, like the blank name and for the same reason: it is
	// the joiner's own input and telling them says nothing about the door.
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}
	for _, name := range []string{
		strings.Repeat("a", 65),             // bounded only by the body cap
		strings.Repeat("\u00e5", 200),       // and bounded by RUNES, not bytes
		"Kim\u001b[31m",                     // an ANSI escape, for anything that logs it
		"Kim\nDM: everyone roll initiative", // a newline, to forge a second line
		"Kim\u202emiK",                      // a BIDI override, so the name reads as somebody else
		"\u200b\u200b\u200b",                // zero-width spaces: passes every other rule, shows nothing
		"\ufeff",                            // a byte-order mark, the same shape
		"\u3164",                            // HANGUL FILLER, the classic invisible-name trick
		"Kim\u0000",                         // a NUL, for anything that is not Go
	} {
		resp, _, _ := f.post(secret, name)
		if resp.StatusCode != http.StatusBadRequest {
			t.Fatalf("display name %q: status %d, want %d", name, resp.StatusCode, http.StatusBadRequest)
		}
	}
	// And an ordinary name with non-ASCII in it still gets in — a bound that
	// only accepts ASCII would lock out most of the people who might play.
	resp, got, _ := f.post(secret, "Åsa Örn")
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("a perfectly ordinary name was refused: %d", resp.StatusCode)
	}
	if got.Name != "Åsa Örn" {
		t.Fatalf("display name came back as %q", got.Name)
	}

	// THE BOUNDARY ITSELF, from both sides. 65 above is refused; exactly the
	// cap must be ACCEPTED, or the limit is quietly 63 and the only person who
	// finds out is the one whose name is that long.
	//
	// The mutation gate caught this gap: > became >= and every test still
	// passed, because nothing sat on the boundary. 64 is written literally
	// rather than read from the constant — this is the external test package,
	// and a test that recomputes the bound from the same expression it is
	// checking cannot catch the bound being wrong.
	resp, _, _ = f.post(secret, strings.Repeat("n", 64))
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("a name of exactly the documented 64-rune limit was refused (%d) — the limit "+
			"is off by one", resp.StatusCode)
	}
}

func TestTwoJoinersGetDistinctIdentities(t *testing.T) {
	// The point of minting per person rather than sharing one credential: two
	// people through the same link are two participants, separately revocable,
	// each keeping their own characters.
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}
	_, a, _ := f.post(secret, "Kim")
	_, b, _ := f.post(secret, "Ada")
	if a.Token == b.Token || a.ParticipantID == b.ParticipantID {
		t.Fatalf("two joiners must get distinct credentials: %+v vs %+v", a, b)
	}
}

func TestRotatingTheLinkRefusesTheOldSecret(t *testing.T) {
	// The property spec §2 calls close to required: a leaked link is closable.
	f := newJoinFixture(t)
	old, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if err := f.ids.SetJoinOpen(true, 100); err != nil {
		t.Fatal(err)
	}
	if resp, _, _ := f.post(old, "Kim"); resp.StatusCode != http.StatusOK {
		t.Fatal("the old secret should work before rotating")
	}
	fresh, err := f.ids.RotateJoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if resp, _, _ := f.post(old, "Mallory"); resp.StatusCode == http.StatusOK {
		t.Fatal("a rotated link must refuse the OLD secret, or rotating protects nobody")
	}
	if resp, _, _ := f.post(fresh, "Ada"); resp.StatusCode != http.StatusOK {
		t.Fatal("the NEW secret must work, or rotating locks the table out")
	}
}

// TestTheDoorStopsAdmittingWhenItsBudgetIsSpent is the END-TO-END wiring, and
// it exists because the registry-level tests cannot see it.
//
// Every identity test injects its own budget and calls JoinAdmits directly. If
// handleJoin kept calling JoinAllows — which still exists, still compiles, and
// still answers the same question WITHOUT spending anything — the cap would be
// perfect in internal/identity and absent from the product. That is the exact
// shape a review caught in this session's previous change: a feature dead in
// production behind a fully green suite.
func TestTheDoorStopsAdmittingWhenItsBudgetIsSpent(t *testing.T) {
	f := newJoinFixture(t)
	secret, err := f.ids.JoinSecret()
	if err != nil {
		t.Fatal(err)
	}
	if err := f.ids.SetJoinOpen(true, 2); err != nil {
		t.Fatal(err)
	}

	for i := range 2 {
		resp, _, body := f.post(secret, fmt.Sprintf("Player %d", i))
		if resp.StatusCode != http.StatusOK {
			t.Fatalf("join %d of 2 returned %d (%q) — the budget was spent early",
				i+1, resp.StatusCode, body)
		}
	}

	spent, _, spentBody := f.post(secret, "One Too Many")
	if spent.StatusCode != http.StatusForbidden {
		t.Fatalf("the third join returned %d, want 403 — an open door mints without limit, "+
			"so a leaked link is bounded only by how fast anyone clicks", spent.StatusCode)
	}

	// The SAME answer as a wrong secret, byte for byte. A prober must not learn
	// which of the three refusals they hit (spec §5) — and the raw body is
	// compared, not the decoded struct, because http.Error writes plain text
	// and two failed decodes are identical whatever the bodies said.
	wrong, _, wrongBody := f.post("not-the-secret", "Prober")
	if wrong.StatusCode != spent.StatusCode || wrongBody != spentBody {
		t.Fatalf("a spent budget answers %d %q and a wrong secret answers %d %q — the "+
			"difference tells a prober the door is real and merely full",
			spent.StatusCode, spentBody, wrong.StatusCode, wrongBody)
	}

	// And it minted exactly the two it admitted.
	if n := f.count("participants"); n != 2 {
		t.Fatalf("%d participants exist, want 2 — the refused join minted one anyway", n)
	}
}

File: client/test/join.test.ts
import "./support/dom"; // see that module: registers once, keeps real fetch/WebSocket

import { test, expect, afterEach } from "bun:test";
import { requestJoin } from "../src/join";
import { renderJoinView, type JoinViewState } from "../src/view/join";

// The shared join link, client side (joining-a-table spec §2, plan J5).
//
// The person on the other end of this has NO credential yet and no way to get
// one except through this form. So the failures matter more than the success:
// a joiner who cannot tell "the DM has not opened the door" from "your name is
// too long" will retype their name until they give up.
//
// The server draws exactly that line for us and it is worth keeping: 400 is
// the joiner's own input and is safe to describe, 403 is deliberately
// identical for a closed door and a wrong secret so a prober learns nothing.

const nativeFetch = globalThis.fetch;
afterEach(() => {
  globalThis.fetch = nativeFetch;
});

interface Seen {
  url: string;
  method: string;
  headers: unknown;
  body: unknown;
}

/** Answer every request with one response, recording what was asked. */
function stubJoin(status: number, body: string, seen: Seen[] = []): Seen[] {
  globalThis.fetch = (async (input: RequestInfo | URL, init?: RequestInit) => {
    seen.push({
      url: String(input),
      method: init?.method ?? "GET",
      headers: init?.headers,
      body: JSON.parse(String(init?.body ?? "null")),
    });
    return new Response(body, { status });
  }) as typeof fetch;
  return seen;
}

test("joining posts the secret and the name, and comes back with a credential", async () => {
  const seen = stubJoin(200, JSON.stringify({ token: "tok-new", participantId: "p-9" }));

  const out = await requestJoin("http://table.example", "s3cret", "Kim");

  expect(out.ok).toBe(true);
  if (out.ok) expect(out.token).toBe("tok-new");
  // The REQUEST, not just the reply. A join that returned the right token
  // while posting the wrong secret would pass a reply-only assertion and fail
  // against a real server.
  expect(seen).toHaveLength(1);
  expect(seen[0]!.url).toBe("http://table.example/join");
  expect(seen[0]!.method).toBe("POST");
  expect(seen[0]!.body).toEqual({ secret: "s3cret", displayName: "Kim" });
  // The content type is declared. Our own handler does not check it, so this
  // costs nothing today — but a body posted without it is at the mercy of
  // whatever proxy or WAF sits between a table and the people joining it.
  expect(seen[0]!.headers).toEqual({ "content-type": "application/json" });
});

test("a refused name is reported as the joiner's own, and without the wire prefix", async () => {
  // 400 is the one refusal the server describes, deliberately: it is the
  // joiner's own input, so saying what is wrong leaks nothing about the door.
  stubJoin(400, "gateway: a display name is required, up to 64 ordinary characters\n");

  const out = await requestJoin("http://table.example", "s3cret", "");

  expect(out.ok).toBe(false);
  if (!out.ok) {
    expect(out.reason).toBe("name");
    // "gateway:" is wire vocabulary. The person reading this typed a name into
    // a box; they did not address a gateway.
    // EXACT, not "contains". The prefix strip, the trim and the fallback all
    // live in one small function, and a containment check cannot see a
    // trailing newline or a half-stripped prefix — which is precisely what
    // the three near-identical regexes it replaced differed by.
    expect(out.message).toBe("a display name is required, up to 64 ordinary characters");
  }

  // A body that carries no wire prefix is passed through UNTOUCHED. Without
  // this, "always strip nine characters" and "strip only when the prefix is
  // there" are the same program.
  stubJoin(400, "that name is already taken at this table");
  const plain = await requestJoin("http://table.example", "s3cret", "Kim");
  expect(plain.ok).toBe(false);
  if (!plain.ok) expect(plain.message).toBe("that name is already taken at this table");

  // A 400 whose body a proxy has eaten still has to say SOMETHING. An empty
  // error box beside a form reads as "this is broken" rather than "fix this".
  stubJoin(400, "");
  const bare = await requestJoin("http://table.example", "s3cret", "");
  expect(bare.ok).toBe(false);
  if (!bare.ok) {
    expect(bare.reason).toBe("name");
    expect(bare.message.length).toBeGreaterThan(0);
  }
});

test("a closed door is never reported as a problem with the name", async () => {
  // The trap this pins: folding 403 into the same "fix your input" message
  // would have somebody retyping their name at a door that is simply shut,
  // with no reason to ever stop.
  stubJoin(403, "gateway: this link is not accepting anyone\n");

  const out = await requestJoin("http://table.example", "wrong", "Kim");

  expect(out.ok).toBe(false);
  if (!out.ok) {
    expect(out.reason).toBe("door");
    // Both halves of the sentence. It has to say what is wrong AND what to do
    // about it: "you cannot come in" with no next step is a dead end for
    // somebody who has no other way to reach this table.
    expect(out.message).toContain("not letting anyone in");
    expect(out.message).toContain("fresh link");
  }
});

test("a success carrying no token is a failure, not a session", async () => {
  // Auth.set THROWS on an empty token, so a 200 with a missing or blank token
  // would take down the boot with an exception instead of showing the joiner
  // anything. The malformed reply has to be caught here, where there is still
  // a form to put a message on.
  // A reply that PARSED but carried nothing usable...
  for (const body of ['{"token":""}', "{}", '{"token":42}']) {
    stubJoin(200, body);
    const out = await requestJoin("http://table.example", "s3cret", "Kim");
    expect(out.ok).toBe(false);
    if (!out.ok) {
      expect(out.reason).toBe("unavailable");
      expect(out.message).toContain("no credential");
    }
  }

  // ...versus one that could not be parsed at all. Two different situations
  // and two different messages: a proxy rewriting the body, or the endpoint
  // itself. Whoever debugs this should not have to guess which.
  stubJoin(200, "not json at all");
  const unreadable = await requestJoin("http://table.example", "s3cret", "Kim");
  expect(unreadable.ok).toBe(false);
  if (!unreadable.ok) {
    expect(unreadable.reason).toBe("unavailable");
    expect(unreadable.message).toContain("could not be read");
  }
});

test("a server error is refused even when it comes with a plausible token", async () => {
  // The status is checked BEFORE the body is trusted. Without that, a 500 or a
  // 502 whose body happens to parse would hand the joiner a credential the
  // server never issued — and every later reload would dial with it and
  // report a connection problem rather than the real failure.
  stubJoin(500, JSON.stringify({ token: "tok-from-an-error-page" }));

  const out = await requestJoin("http://table.example", "s3cret", "Kim");

  expect(out.ok).toBe(false);
  if (!out.ok) {
    expect(out.reason).toBe("unavailable");
    expect(out.message).toContain("Could not reach the table");
  }
});

test("a network failure is an answer, not an unhandled rejection", async () => {
  // `as unknown as` rather than a direct assertion: `typeof fetch` carries
  // `preconnect`, and TS2352 fails the typecheck without the widening step.
  globalThis.fetch = (async () => {
    throw new TypeError("Failed to fetch");
  }) as unknown as typeof fetch;

  const out = await requestJoin("http://table.example", "s3cret", "Kim");

  expect(out.ok).toBe(false);
  if (!out.ok) {
    expect(out.reason).toBe("unavailable");
    expect(out.message).toContain("Check your connection");
  }
});

// --- the view -----------------------------------------------------------

function root(): HTMLElement {
  const el = document.createElement("div");
  document.body.replaceChildren(el);
  return el;
}

function state(over: Partial<JoinViewState> = {}): JoinViewState {
  return { name: "", busy: false, error: "", ...over };
}

test("the form asks for a name and hands back what was typed", () => {
  const r = root();
  let submitted = "";
  let typed = "";
  renderJoinView(r, state(), {
    onName: (v) => {
      typed = v;
    },
    onSubmit: () => {
      submitted = typed;
    },
  });

  const input = r.querySelector<HTMLInputElement>('[data-field="join-name"]')!;
  input.value = "Kim";
  input.dispatchEvent(new Event("input"));
  const btn = r.querySelector<HTMLButtonElement>('[data-action="join"]')!;
  expect(btn.textContent).toBe("Join");
  expect(btn.disabled).toBe(false);
  btn.click();

  expect(submitted).toBe("Kim");
  // The form names itself and names the field. A join screen with no heading
  // and no label is a bare box on a blank page.
  expect(r.querySelector(".join h1")!.textContent).toBe("Join the table");
  const label = r.querySelector<HTMLLabelElement>(".join label")!;
  expect(label.textContent!.length).toBeGreaterThan(0);
  expect(label.htmlFor).toBe(input.id);
});

test("Enter submits, because nobody reaches for the mouse after typing their name", () => {
  const r = root();
  let submits = 0;
  renderJoinView(r, state({ name: "Kim" }), { onName: () => {}, onSubmit: () => submits++ });

  const input = r.querySelector<HTMLInputElement>('[data-field="join-name"]')!;
  input.dispatchEvent(new KeyboardEvent("keydown", { key: "a" }));
  expect(submits).toBe(0); // an ordinary keystroke is not a submit
  input.dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" }));
  expect(submits).toBe(1);
});

test("while the join is in flight the button cannot be pressed again", () => {
  // Every press mints a PARTICIPANT. A double-click during a slow round trip
  // would put the same person at the table twice, holding two credentials,
  // one of which nothing will ever revoke because nobody knows it exists.
  const r = root();
  let submits = 0;
  renderJoinView(r, state({ busy: true }), { onName: () => {}, onSubmit: () => submits++ });

  const btn = r.querySelector<HTMLButtonElement>('[data-action="join"]')!;
  expect(btn.disabled).toBe(true);
  // And it SAYS so. A button that is merely inert looks like a dead page;
  // this is the only feedback there is while the round trip is in flight.
  expect(btn.textContent).toBe("Joining\u2026");
  btn.click();
  expect(submits).toBe(0);
});

test("a refusal is shown, and the name the person already typed survives it", () => {
  // The repaint that carries the error also rebuilds the input. Losing the
  // value there makes every failed attempt cost the joiner their typing,
  // which is worst exactly when the door is shut and they will try again.
  const r = root();
  renderJoinView(r, state({ name: "Kim", error: "The DM has not opened this table yet." }), {
    onName: () => {},
    onSubmit: () => {},
  });

  expect(r.querySelector(".error")!.textContent).toBe("The DM has not opened this table yet.");
  expect(r.querySelector<HTMLInputElement>('[data-field="join-name"]')!.value).toBe("Kim");
});

test("with nothing wrong there is no error element to read", () => {
  // An empty error node still occupies the page and reads as a blank warning.
  const r = root();
  renderJoinView(r, state(), { onName: () => {}, onSubmit: () => {} });
  expect(r.querySelector(".error")).toBeNull();
});
